src/common.c

```diff
@@ -1602,19 +1602,22 @@ static void ofi_pollfds_do_del(struct ofi_pollfds *pfds,
 	if (ofi_poll_fairness && ctx->hot_index >= 0)
 		ofi_pollfds_coolfd(pfds, item->fd);
 
+	last_pfd = &pfds->fds[pfds->nfds - 1];
+	last_ctx = ofi_pollfds_get_ctx(pfds, last_pfd->fd);
+	assert(last_ctx);
 	if (ctx->index < pfds->nfds - 1) {
-		swap_pfd = &pfds->fds[pfds->nfds - 1];
-		swap_ctx = ofi_pollfds_get_ctx(pfds, swap_pfd->fd);
-		assert(swap_ctx);
-		swap_ctx->index = ctx->index;
-		pfds->fds[swap_ctx->index] = *swap_pfd;
-
-		swap_pfd->fd = INVALID_SOCKET;
-		swap_pfd->events = 0;
-		swap_pfd->revents = 0;
-	}
+		last_ctx->index = ctx->index;
+		pfds->ctx[ctx->index] = *last_ctx;
```

Comment:
This looks like the relevant change.  Can you describe the problem in more detail in the commit message?  The code below the if statement mostly appears preventative.

Reply:
Preventative, but required. When adding an element there is an assert that the element it is about to use is not already in use `assert(ctx->index == -1)` Tripping that assert is what eventually led me here.

In addition to not copying the ofi_pollfds_ctx from the last position to the position being deleted, the setting of `ctx->index = -1` was not correct when the if statement had been taken, because it would obliterate the very element that it had just moved.

I'll make the 0 a NULL and update the commit message to reflect what I just described.

Reply:
On linux, this works as follows.  The fds[] is a compact array of file descriptors.  The ctx[] is a sparse array of context values.  The value of any fd is used as a direct index into the ctx[] array.  That provides a mapping where the fd is stored in the fds[] array.  So, the ctx[] structures are never swapped, only the fds[] entry changes.  So, the changes below where last_ctx is modified would break linux as there's no swapping occurring.  The change you added to line 1610 ends up being unnecessary.  I think we may need to pull ofi_pollfds_do_del() back out into OS specific implementations, or maybe introduce a new OS specific swap function.

Reply:
Hmm...Guess I was a little too focused on windows and didn't appreciate the amount of nuance here.

Reply:
That's okay, I was focused on linux mostly when I wrote the code.  :)
But I think that for linux, we need to swap the fds[] entries only.  For windows, we either need to swap both or update the windows code to allow for unused ctx[] entries.  The latter might be possible by updating ofi_pollfds_alloc_ctx() to return the first unused ctx[] entry.